src/topology/discrete_quotient.lean

```diff
@@ -0,0 +1,233 @@
+2. `discrete_quotient.exists_of_compat` which states that when `X` is compact, then any
+  system of elements of `Q` as `Q : discrete_quotient X` varies, which is compatible with
+  respect to `discrete_quotient.map`, must arise from some element of `X`.
+
+## Remarks
+The constructions in this file will be used to show that any profinite space is a limit
+of finite discrete spaces.
+-/
+
+variables (X : Type*) [topological_space X]
+
+/-- The type of discrete quotients of a topological space. -/
+@[ext]
+structure discrete_quotient :=
+(rel : X → X → Prop)
+(equiv : equivalence rel)
+(clopen : ∀ x, is_clopen (set_of (rel x)))
+
+namespace discrete_quotient
+
+variables {X} (S : discrete_quotient X)
+
+/-- Construct a discrete quotient from a clopen set. -/
+def of_clopen {A : set X} (h : is_clopen A) : discrete_quotient X :=
+{ rel := λ x y, x ∈ A ∧ y ∈ A ∨ x ∉ A ∧ y ∉ A,
+  equiv := ⟨by tauto!, by tauto!, by tauto!⟩,
+  clopen := begin
+    intros x,
+    by_cases hx : x ∈ A,
+    { apply is_clopen_union,
+      { convert h,
+        ext,
+        exact ⟨λ i, i.2, λ i, ⟨hx,i⟩⟩ },
+      { convert is_clopen_empty,
+        tidy } },
+    { apply is_clopen_union,
+      { convert is_clopen_empty,
+        tidy },
+      { convert is_clopen_compl h,
+        ext,
+        exact ⟨λ i, i.2, λ i, ⟨hx, i⟩⟩ } },
+  end}
+
+lemma refl : ∀ x : X, S.rel x x := S.equiv.1
+lemma symm : ∀ x y : X, S.rel x y → S.rel y x := S.equiv.2.1
+lemma trans : ∀ x y z : X, S.rel x y → S.rel y z → S.rel x z := S.equiv.2.2
+
+/-- The setoid whose quotient yields the discrete quotient. -/
+def setoid : setoid X := ⟨S.rel, S.equiv⟩
+
+instance : has_coe_to_sort (discrete_quotient X) :=
+⟨Type*, λ S, quotient S.setoid⟩
+
+instance : topological_space S := ⊥
+
+/-- The projection from `X` to the given discrete quotient. -/
+def proj : X → S := quotient.mk'
+
+lemma proj_surjective : function.surjective S.proj := quotient.surjective_quotient_mk'
+
+lemma fiber_eq (x : X) : S.proj ⁻¹' {S.proj x} = set_of (S.rel x) :=
+begin
+  ext1 y,
+  simp only [set.mem_preimage, set.mem_singleton_iff, quotient.eq',
+    discrete_quotient.proj.equations._eqn_1, set.mem_set_of_eq],
+  exact ⟨λ h, S.symm _ _ h, λ h, S.symm _ _ h⟩,
+end
+
+lemma proj_is_locally_constant : is_locally_constant S.proj :=
+begin
+   rw (is_locally_constant.tfae S.proj).out 0 3,
+   intros x,
+   rcases S.proj_surjective x with ⟨x,rfl⟩,
+   simp [fiber_eq, (S.clopen x).1],
+end
+
+lemma proj_continuous : continuous S.proj :=
+is_locally_constant.continuous $ proj_is_locally_constant _
+
+lemma fiber_closed (A : set S) : is_closed (S.proj ⁻¹' A) :=
+is_closed.preimage S.proj_continuous ⟨trivial⟩
+
+lemma fiber_open (A : set S) : is_open (S.proj ⁻¹' A) :=
+is_open.preimage S.proj_continuous trivial
+
+lemma fiber_clopen (A : set S) : is_clopen (S.proj ⁻¹' A) := ⟨fiber_open _ _, fiber_closed _ _⟩
+
+/-- Pull back a discrete quotient along a continuous map. -/
+def pullback {Y : Type*} [topological_space Y] {f : Y → X}
```

Comment:
Should this be called `comap`? That's what mathlib uses in the unbundled world.

Reply:
I don't mind renaming this `comap`, but then we should change the name of what's currently called `discrete_quotient.map`. Any suggestions for a better name for `map`?

Reply:
maybe `of_le` or `hom_of_le`?

Reply:
Okay, I went with `of_le` (and changed `pullback` -> `comap`)